backend/database/versions/26_e08b6b67f076_make_collection_id_pkey.py

```diff
@@ -0,0 +1,126 @@
+    op.create_unique_constraint(op.f("_geneset_name__collection_uc"), "geneset", ["name", "collection_id"])
+
+    # Add project.revision_of
+    op.add_column("project", sa.Column("revision_of", sa.String(), nullable=True))
+
+    # Drop project composite primary key; add project.id as primary key
+    op.drop_constraint("project_pkey", "project")
+    op.create_primary_key(op.f("pk_project"), "project", ["id"])
+
+    # Add foreign keys for project.id
+    op.create_foreign_key(op.f("fk_dataset_collection_id_project"), "dataset", "project", ["collection_id"], ["id"])
+    op.create_foreign_key(op.f("fk_geneset_collection_id_project"), "geneset", "project", ["collection_id"], ["id"])
+    op.create_foreign_key(op.f("fk_project_revision_of_project"), "project", "project", ["revision_of"], ["id"])
+    op.create_foreign_key(
+        op.f("fk_project_link_collection_id_project"), "project_link", "project", ["collection_id"], ["id"]
+    )
+
+    # Check constraint: published Collections must not be a revision_of another Collection
+    op.create_check_constraint(
+        op.f("ck_project_private_revision"), "project", "revision_of IS NULL OR visibility = 'PRIVATE'"
+    )
+
+
+def downgrade():
+    def remake_collection_id_and_collection_visibility_composite_foreign_key(table: str):
+        op.add_column(
+            table,
+            sa.Column(
+                "collection_visibility",
+                postgresql.ENUM("PRIVATE", "PUBLIC", name="collectionvisibility"),
+                autoincrement=False,
+            ),
+        )
+        op.execute(
+            f"""
+            UPDATE {table}
+                SET collection_visibility = (
+                    SELECT visibility
+                    FROM project
+                    WHERE project.id = {table}.collection_id
+                )
+            """
+        )
+        op.alter_column(table, "collection_visibility", nullable=False)
+        op.create_foreign_key(
+            f"{table}_collection_id_fkey",
+            table,
+            "project",
+            ["collection_id", "collection_visibility"],
+            ["id", "visibility"],
+        )
+
+    # Drop foreign keys for project.id
+    op.drop_constraint(op.f("fk_project_link_collection_id_project"), "project_link", type_="foreignkey")
+    op.drop_constraint(op.f("fk_project_revision_of_project"), "project", type_="foreignkey")
+    op.drop_constraint(op.f("fk_geneset_collection_id_project"), "geneset", type_="foreignkey")
+    op.drop_constraint(op.f("fk_dataset_collection_id_project"), "dataset", type_="foreignkey")
+
+    # Drop project.id as primary key
+    op.drop_constraint(op.f("pk_project"), "project")
+
+    # Safety: save current project.id in extra column as backup -- BREAKS MIGRATION FLOW
+    op.add_column("project", sa.Column("id_bk", sa.String()))
+    op.execute("UPDATE project SET id_bk = id")
+    op.alter_column("project", "id_bk", nullable=False)
```

Comment:
Is this step of backing up the unique id's necessary? I've added it in case of a rogue/accidental migration downgrade.

Reply:
TLDR: honorable engineering, but probably not worth the effort.

If we have to downgrade a prod migration, I would say all bets are off as to the permanence of any private collection URLs users may have been acquired by users. And I think this would only affect private revisions that were curated after the "up" migration (since no revisions can be in progress at that point).  So it's a very limited blast radius for this problem, even if it takes a couple of days to catch a problem.

Reply:
Ok, will remove.